pkg/services/tracer/middleware.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"runtime/debug"
+	"strconv"
+	"strings"
+
+	"github.com/urfave/negroni"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+)
+
+func Tracer(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if isHealthRoute(r) || isSwaggerRoute(r) {
+			next.ServeHTTP(w, r)
+			return
+		}
+		span, traceCtx := createSpanAndContext(r, opentracing.GlobalTracer())
+		defer span.Finish()
+		defer setSpanIfPanic(span, w)
+		status := setHTTPSpans(traceCtx, w, r, span, next)
+		setSpanErrorIfStatus(status, span)
+	})
+}
+
+func createSpanAndContext(r *http.Request, tr opentracing.Tracer) (opentracing.Span, context.Context) {
+	operationName, serverSpanCtx := getContextFromHeader(r, tr)
+	span, traceCtx := opentracing.StartSpanFromContext(r.Context(),
+		operationName, ext.RPCServerOption(serverSpanCtx))
+	return span, traceCtx
+}
+
+func setSpanIfPanic(span opentracing.Span, w http.ResponseWriter) {
+	if err := recover(); err != nil {
+		ext.HTTPStatusCode.Set(span, http.StatusInternalServerError)
+		SetSpanError(span, errors.New(fmt.Sprint(err)))
+		span.SetTag("error.type", "panic")
+		span.LogKV(
+			"event", "error",
+			"error.kind", "panic",
+			"message", err,
+			"stack", string(debug.Stack()),
+		)
+		span.Finish()
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func setHTTPSpans(traceCtx context.Context, w http.ResponseWriter, r *http.Request,
+	span opentracing.Span, next http.Handler) int {
+	ext.HTTPMethod.Set(span, r.Method)
+	ext.HTTPUrl.Set(span, r.URL.Path)
+
+	resourceName := r.URL.Path
+	resourceName = r.Method + " " + resourceName
+	span.SetTag("resource.name", resourceName)
+
+	ww := negroni.NewResponseWriter(w)
```

Comment:
The function `NewWrapResponseWriter` is also provided by `go-chi` on package `github.com/go-chi/chi/middleware`

Reply:
when testing https://github.com/go-chi/httptracer wich uses `NewWrapResponseWriter` it was panicking on this function, so i wrote the middleware based on them with another wrapper